databuilder/query_model.py

```diff
@@ -1,452 +1,356 @@
 
-    def __le__(self, other):
-        return self._get_comparator("__le__", other)
 
-    def __eq__(self, other):
-        return self._get_comparator("__eq__", other)
+class BinaryFunction(Column):
+    lhs: Union[Column, Value]
+    rhs: Union[Column, Value]
 
-    def __ne__(self, other):
-        return self._get_comparator("__ne__", other)
 
-    def __and__(self, other):
-        other = self._other_as_comparator(other)
-        return boolean_comparator(self) & other
+# All available functions (using a class as a namespace)
+class Function:
 
-    def __or__(self, other):
-        other = self._other_as_comparator(other)
-        return boolean_comparator(self) | other
+    # Comparison
+    class EQ(BinaryFunction):
+        ...
 
-    def __invert__(self):
-        return boolean_comparator(self, negated=True)
+    class NE(BinaryFunction):
+        ...
 
-    def __hash__(self):
-        return id(self)
+    class LT(BinaryFunction):
+        ...
 
+    class LE(BinaryFunction):
+        ...
 
-@dataclass(frozen=True, eq=False, order=False)
-class ValueFromRow(Value):
-    source: Any
-    column: Any
+    class GT(BinaryFunction):
+        ...
 
-    def _get_referenced_nodes(self):
-        return (self.source,)
+    class GE(BinaryFunction):
+        ...
 
+    # Boolean
+    class And(BinaryFunction):
+        ...
 
-@dataclass(frozen=True, eq=False, order=False)
-class ValueFromAggregate(Value):
-    source: RowFromAggregate
-    column: Any
+    class Or(BinaryFunction):
+        ...
 
-    def _get_referenced_nodes(self):
-        return (self.source,)
+    class Not(UnaryFunction):
+        ...
 
+    # Null handling
+    class IsNull(UnaryFunction):
+        ...
 
-def categorise(mapping, default=None):
-    mapping = {
-        key: boolean_comparator(value) if isinstance(value, Value) else value
-        for key, value in mapping.items()
-    }
-    return ValueFromCategory(mapping, default)
+    # Arithmetic
+    class Add(BinaryFunction):
+        ...
 
+    class Subtract(BinaryFunction):
+        ...
 
-@dataclass(frozen=True, eq=False, order=False)
-class ValueFromCategory(Value):
-    definitions: dict
-    default: str | int | float | None
+    # Dates
+    class RoundToFirstOfMonth(UnaryFunction):
+        ...
 
-    def _get_referenced_nodes(self):
-        nodes = ()
-        for value in self.definitions.values():
-            assert isinstance(value, QueryNode)
-            nodes += (value,)
-        return nodes
+    class RoundToFirstOfYear(UnaryFunction):
+        ...
 
+    class DateAdd(BinaryFunction):
+        ...
 
-@dataclass(frozen=True)
-class Codelist(QueryNode):
-    codes: tuple
-    system: str
-    has_categories: bool = False
+    class DateSubtract(BinaryFunction):
+        ...
 
-    def _get_referenced_nodes(self):
-        return ()
+    class DateDifference(Column):
+        start: Union[Column, datetime.date]
+        end: Union[Column, datetime.date]
+        units: str
```

Comment:
This might just be a case of the function being badly named. This is a function which returns an int and which we use for calculating someone's age on a given date by using it with the `years` unit. I think it also supports `months`, `weeks` and `days`. We don't have date delta as a type in itself at the moment.

Reply:
Oh okay. I guess that does the job, doesn't it? No point agonizing over it at this point, I think.